ARCs/arc-0056.md

```diff
@@ -0,0 +1,343 @@
+
+```ts
+/** Describes the entire contract. This interface is an extension of the interface described in ARC-4 */
+interface Contract {
+  /** The ARCs used and/or supported by this contract. All contracts implicity support ARC4 and ARC56 */
+  arcs: number[];
+  /** A user-friendly name for the contract */
+  name: string;
+  /** Optional, user-friendly description for the interface */
+  desc?: string;
+  /**
+   * Optional object listing the contract instances across different networks
+   */
+  networks?: {
+    /**
+     * The key is the base64 genesis hash of the network, and the value contains
+     * information about the deployed contract in the network indicated by the
+     * key. A key containing the human-readable name of the network MAY be
+     * included, but the corresponding genesis hash key MUST also be defined
+     */
+    [network: string]: {
+      /** The app ID of the deployed contract in this network */
+      appID: number;
+    };
+  };
+  /** Named structs use by the application */
+  structs: { [structName: StructName]: StructFields };
+  /** All of the methods that the contract implements */
+  methods: Method[];
+  state: {
+    /** Defines the values that should be used for GlobalNumUint, GlobalNumByteSlice, LocalNumUint, and LocalNumByteSlice when creating the application  */
+    schema: {
+      global: {
+        ints: number;
+        bytes: number;
+      };
+      local: {
+        ints: number;
+        bytes: number;
+      };
+    };
+    /** Mapping of human-readable names to StorageKey objects */
+    keys: {
+      global: { [name: string]: StorageKey };
+      local: { [name: string]: StorageKey };
+      box: { [name: string]: StorageKey };
+    };
+    /** Mapping of human-readable names to StorageMap objects */
+    maps: {
+      global: { [name: string]: StorageMap };
+      local: { [name: string]: StorageMap };
+      box: { [name: string]: StorageMap };
+    };
+  };
+  /** Supported bare actions for the contract. An action is a combination of call/create and an OnComplete */
+  bareActions: {
+    /** OnCompletes this method allows when appID === 0 */
+    create: ("NoOp" | "OptIn" | "DeleteApplication")[];
+    /** OnCompletes this method allows when appID !== 0 */
+    call: (
+      | "NoOp"
+      | "OptIn"
+      | "CloseOut"
+      | "ClearState"
+      | "UpdateApplication"
+      | "DeleteApplication"
+    )[];
+  };
+  /** Information about the TEAL */
+  sourceInfo?: SourceInfo[];
+  /** The pre-compiled TEAL that may contain template variables. MUST be omitted if included as part of ARC23, but otherwise MUST be defined. */
+  source?: {
+    /** The approval program */
+    approval: string;
+    /** The clear program */
+    clear: string;
+  };
+  /** ARC-28 events that MAY be emitted by this contract */
+  events?: Array<Event>;
+  /** A mapping of template variable names as they appear in the teal (not including TMPL_ prefix) and their respecive types */
```

Comment:
There doesn't seem to be any harm if we include the `TMPL_` prefix? The less assumptions (within the bounds of practicality) the better, no?

Reply:
But if we use it for code generation then you’d have to strip the TMPL so we have to cater for it somewhere?

Reply:
If certain tooling expects that as a prefix, that seems like the place to check for and strip that prefix, yeah?

Reply:
*shrug* could go either way,  I guess it's fair that this could be more general purpose and then the generators are clever enough to detect and strip TMPL_ when emitting code...

Reply:
My rationale omitting the TMPL_ here and forcing tooling to use it is to make it easier to implement safely. For example, if ARC56 allowed anything as the variable name and the tooling didn't have to add `TMPL_`, it would need to check all the variable names against all opcodes. Otherwise an ARC56 file could have a template variable named `byte` which subsequently overwrites all `byte` opcodes which is not the intended effect of template variables.